File: src/algorithms/dynamic-programming/subsequence/minimum-jumps/sources/minimum-jumps.rs
// Minimum Jumps tabulation — build DP table iteratively from base case

fn minimum_jumps(jumps: &[usize]) -> i64 {
    // @step:initialize
    let array_length = jumps.len(); // @step:initialize
    if array_length == 0 {
        return 0; // @step:initialize
    }
    let mut dp_table = vec![i64::MAX; array_length]; // @step:initialize,fill-table
    dp_table[0] = 0; // @step:fill-table
    // For each position, check all prior positions that can reach it
    for target_index in 1..array_length {
        // @step:compute-cell
        for source_index in 0..target_index {
            // @step:read-cache
            if dp_table[source_index] != i64::MAX
                && source_index + jumps[source_index] >= target_index
            {
                // @step:read-cache
                let candidate = dp_table[source_index] + 1;
                if candidate < dp_table[target_index] {
                    dp_table[target_index] = candidate; // @step:compute-cell,read-cache
                }
            }
        }
    }
    if dp_table[array_length - 1] == i64::MAX {
        -1
    } else {
        dp_table[array_length - 1] // @step:complete
    }
}

fn main() {
    let jumps = vec![2, 3, 1, 1, 4];
    let result = minimum_jumps(&jumps);
    println!("Minimum jumps for {:?}: {}", jumps, result);
}

File: src/algorithms/dynamic-programming/subsequence/minimum-jumps/sources/MinimumJumps.cpp
// Minimum Jumps tabulation — build DP table iteratively from base case

#include <iostream>
#include <vector>
#include <climits>
#include <algorithm>

int minimumJumps(const std::vector<int>& jumps) {
    // @step:initialize
    int arrayLength = jumps.size(); // @step:initialize
    if (arrayLength == 0) return 0; // @step:initialize
    std::vector<int> dpTable(arrayLength, INT_MAX); // @step:initialize,fill-table
    dpTable[0] = 0; // @step:fill-table
    // For each position, check all prior positions that can reach it
    for (int targetIndex = 1; targetIndex < arrayLength; targetIndex++) {
        // @step:compute-cell
        for (int sourceIndex = 0; sourceIndex < targetIndex; sourceIndex++) {
            // @step:read-cache
            if (dpTable[sourceIndex] != INT_MAX
                && sourceIndex + jumps[sourceIndex] >= targetIndex) {
                // @step:read-cache
                int candidate = dpTable[sourceIndex] + 1;
                if (candidate < dpTable[targetIndex]) {
                    dpTable[targetIndex] = candidate; // @step:compute-cell,read-cache
                }
            }
        }
    }
    return dpTable[arrayLength - 1] == INT_MAX ? -1 : dpTable[arrayLength - 1]; // @step:complete
}

#ifndef TESTING
int main() {
    std::vector<int> jumps = {2, 3, 1, 1, 4};
    int result = minimumJumps(jumps);
    std::cout << "Minimum jumps: " << result << std::endl;
    return 0;
}
#endif
